Spigot-Server-Patches/0332-Optimize-World-search-radius-functions.patch

```diff
@@ -0,0 +1,215 @@
+From 3ec6476aeadd6d74b334f0d5e7b1ee14bdc0d3c7 Mon Sep 17 00:00:00 2001
+From: egg82 <phantom_zero@ymail.com>
+Date: Tue, 7 Aug 2018 17:55:35 -0600
+Subject: [PATCH] Optimize World search/radius functions
+
+Specifically fixes #1289
+
+Affected functions:
+World.a(double d0, double d1, double d2, double d3, Predicate<Entity> predicate)
+World.isPlayerNearby(double d1, double d2, double d3, double searchRadius)
+World.a(double d0, double d1, double d2, double d3, double d4, @Nullable Function<EntityHuman, Double> function, @Nullable Predicate<EntityHuman> predicate)
+
+Renamed variables in functions and headers, and completely replaced said functions
+The rename shouldn't affect anything but readability
+
+The new code utilizes CraftWorld.getNearbyEntities for later optimization in a future patch
+It shouldn't affect previous behavior and in fact makes attempts at heading off edge-cases early
+
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 004c3ec4..1307e503 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -24,6 +24,7 @@ import com.google.common.collect.Maps;
+ import java.util.HashMap; // Paper
+ import java.util.Map;
+ import org.bukkit.Bukkit;
++import org.bukkit.Location; // Paper
+ import org.bukkit.block.BlockState;
+ import org.bukkit.craftbukkit.CraftServer;
+ import org.bukkit.craftbukkit.CraftWorld;
+@@ -2575,7 +2576,8 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         return i;
+     }
+ 
+-    @Nullable
++    // Paper start
++    /*@Nullable
+     public EntityHuman a(double d0, double d1, double d2, double d3, Predicate<Entity> predicate) {
+         double d4 = -1.0D;
+         EntityHuman entityhuman = null;
+@@ -2599,9 +2601,47 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         }
+ 
+         return entityhuman;
++    }*/
++    @Nullable
++    public EntityHuman a(double centerX, double centerY, double centerZ, double searchRadius, Predicate<Entity> predicate) {
++        if (searchRadius < 0.0d) {
++            // Keeping with previous behavior
```

Comment:
this is wrong. This is not what the previous code did...

I guess seeing this, we will need to keep the old logic still, but only for d3 < 0 scenario.

so if d3 < 0, use old logic, else use new logic

Reply:
Hmm, I swear I double-checked against vanilla behavior to make sure they were at least similar before PRing- I might have missed something here, though. I'll triple-check again and run some more tests.

I ran through cases like d3 < 0 in my head and figured out it would (eventually) simply pull the last player in the players list. I basically short-circuited that logic and skipped all of the checks in that scenario (which happens a surprising amount, weirdly enough. Who'da thunkit that things like "d3 < 0" wouldn't actually be a super edge-case scenario)

Reply:
Had a look at the code again;

The current implementation iterates all possible players in the world and then tests distance against d3 (searchRadius) squared. If d3/searchRadius < 0 AND the player is closer than the previous player it'll use that value.

So, you're right- I missed the second part of that. The effect wouldn't be "pulling the last player from the list" - it would be "pulling the closest player to the origin regardless of distance". I'll fix that.

Reply:
still waiting on fix

Reply:
Good on Github for (not) telling me about this comment. It should be fixed with the latest changes.

Reply:
Sorry, further clarification: I fixed that particular issue in the last round of modifications. The only function that uses that behavior is `boolean isPlayerNearby(double centerX, double centerY, double centerZ, double searchRadius)` because searchRadius < 0 acts as an infinite search radius, and thus will return true if any player is in the world at all, or false if not. That's the only function like that left in the new code. The rest of them have had that logic changed to return the "closest player regardless of distance" instead of "the last player that joined the world" as it did (erroneously) previously.